File: project/jimmer-core/src/test/java/org/babyfish/jimmer/dto/BookDtoTest.java
package org.babyfish.jimmer.dto;

import org.babyfish.jimmer.model.Book;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

public class BookDtoTest {

    @Test
    public void testWithName() {
        BookDto bookDto = new BookDto(
                "SQL in Action",
                "1",
                "32",
                new BookDto.TargetOf_store(
                        "TURING"
                ),
                Arrays.asList(
                        new BookDto.TargetOf_authors("Jim"),
                        new BookDto.TargetOf_authors("Linda")
                )
        );
        Book book = bookDto.toEntity();
        assertContent(
                "{" +
                        "--->\"name\":\"SQL in Action\"," +
                        "--->\"store\":{\"name\":\"TURING\"}," +
                        "--->\"price\":32,\"authors\":[" +
                        "--->--->{\"name\":\"Jim\"}," +
                        "--->--->{\"name\":\"Linda\"}" +
                        "--->]" +
                        "}",
                book
        );
    }

    @Test
    public void testWithoutName() {
        BookDto bookDto = new BookDto(
                null,
                "1",
                "32",
                new BookDto.TargetOf_store(
                        "TURING"
                ),
                Arrays.asList(
                        new BookDto.TargetOf_authors("Jim"),
                        new BookDto.TargetOf_authors("Linda")
                )
        );
        Book book = bookDto.toEntity();
        assertContent(
                "{" +
                        "--->\"store\":{\"name\":\"TURING\"}," +
                        "--->\"price\":32,\"authors\":[" +
                        "--->--->{\"name\":\"Jim\"}," +
                        "--->--->{\"name\":\"Linda\"}" +
                        "--->]" +
                        "}",
                book
        );
    }

    private static void assertContent(String content, Object o) {
        Assertions.assertEquals(content.replace("--->", ""), o.toString());
    }
}

File: project/jimmer-core-kotlin/src/test/kotlin/org/babyfish/jimmer/kt/dto/BookDtoTest.kt
package org.babyfish.jimmer.kt.dto

import org.junit.Test
import kotlin.test.expect

class BookDtoTest {

    @Test
    fun test() {
        val bookDto = BookDto(
            "SQL in Action",
            "1",
            "65",
            BookDto.TargetOfStore("TRUING"),
            listOf(BookDto.TargetOfAuthors("Jim", "Cook"), BookDto.TargetOfAuthors("Linda", "White"))
        )
        val book = bookDto.toEntity()
        assertContents(
            """{
                |--->"name":"SQL in Action",
                |--->"edition":1,
                |--->"price":65,
                |--->"store":{"name":"TRUING"},
                |--->"authors":[
                |--->--->{"firstName":"Jim","lastName":"Cook"},
                |--->--->{"firstName":"Linda","lastName":"White"}
                |--->]
                |}""".trimMargin(),
            book
        )
    }

    companion object {
        fun assertContents(content: String, o: Any) {
            expect(
                content
                    .replace("\r", "")
                    .replace("\n", "")
                    .replace("--->", "")
            ) {
                o.toString()
            }
        }
    }
}